### PR TITLE
[Refactor] Faster and more generic multi-agent nets

### PR DESCRIPTION
**Pull request by vmoens**
cc @matteobettini  @kfu02

TODO:
- [x] Account for non-initialized params
- [x] Have TD.from_modules work with lazy params (the issue here being that the list of parameters will change between before and after the first call is made).

### CODE REVIEW

File: torchrl/modules/models/multiagent.py [hunk 1]
```diff
@@ -150,87 +243,52 @@ def __init__(
...
-                f" but got {output.shape}"
-            )
-
-        return output
+            inputs = inputs.flatten(-2, -1)
+        return inputs
+
+    def _build_single_net(self, *, device, **kwargs):
+        n_agent_inputs = self.n_agent_inputs
+        if self.centralised and n_agent_inputs is not None:
```
Comment (matteobettini) [inline torchrl/modules/models/multiagent.py:278]: Super cool that we are allowing `n_agent_inputs` to be `None` and thus have lazy modules.

We need to document this in the docstring and in the type hint

File: torchrl/modules/models/multiagent.py [hunk 2]
```diff
@@ -2,20 +2,113 @@
...
+        self.share_params = share_params
+        self.centralised = centralised
+        self.agent_dim = agent_dim
+
+        agent_networks = [
+            self._build_single_net(**kwargs)
+            for _ in range(self.n_agents if not self.share_params else 1)
+        ]
+        kwargs["device"] = "meta"
+        self.__dict__["_empty_net"] = self._build_single_net(**kwargs)
```
Comment (matteobettini) [inline torchrl/modules/models/multiagent.py:46]: Just curious the resason for these 2 lines. I don't understand them

**Review by matteobettini — APPROVED** [review]
This is amazing. Thanks so much for this! Can't wait to see the perf gains.

Before merging, we should run a comparision on main and on this PR on one of the multiagent example scripts (e.g., mappo_ippo.py) to test that the reward is the same and gather data for the performance differences.

**Comment by vmoens** [comment]
It is faster, we ran multiple benchmarks on this.
It is faster in low regimes if you consider the backward pass (which is much slower when you build multiple graphs) and when you populate your optimizers with many more params (since you'll be calling `step` and `zero` grad on many more tensors with ops executed in python loops).

File: torchrl/modules/models/multiagent.py (hunk 2 above)
Comment (vmoens) [reply torchrl/modules/models/multiagent.py:58]: To execute the functional module we need an "empty" copy (with meta device - ie no device - ie no data). We also want that copy not to be registered within the list of modules to avoid having those empty parameters pop up when we cann `module.parameters()`.

**Review by kfu02 — APPROVED** [review]
This is cool! Thank you so much! 

To follow up on your comment [here](https://github.com/pytorch/rl/pull/1497#issuecomment-1949335593), what do you anticipate adding an RNN will require in addition to this? Is the issue that the hidden state must be initialized first? I don't quite understand what "non initialized lazy params" means or how that creates an issue with RNNs.

**Comment by vmoens** [comment]
@kfu02 @matteobettini for context: the problem with unitilialized params is that if you have modules with lazy, non initialized params it is usually assumed that you can pass them to an optimizer and the optmizer will know that it must wait until they are initialized to do smth with them.

Currently this isn't supported by `from_modules` which will only create dense params. We could create lazy params but I'm not super duper sure I see how that will work with vmap...

**Comment by matteobettini** [comment]
> It is faster, we ran multiple benchmarks on this.

That is cool! I would still run the full multiagent training script to check those 2 things



> @kfu02 @matteobettini for context: the problem with unitilialized params is that if you have modules with lazy, non initialized params it is usually assumed that you can pass them to an optimizer and the optmizer will know that it must wait until they are initialized to do smth with them.


Since lazy modules was not a feature of these classes in the first place, why do we just not leave it that way?

i personally would prefer this to a complex solution to support it

**Comment by vmoens** [comment]
It was a feature of CNN so I just made it uniform

File: torchrl/modules/models/multiagent.py [hunk 3]
```diff
@@ -2,20 +2,146 @@
...
+
+    def _make_params(self, agent_networks):
+        if self.share_params:
+            self.params = TensorDict.from_module(agent_networks[0], as_module=True)
+        else:
+            self.params = TensorDict.from_modules(*agent_networks, as_module=True)
+
+    @abc.abstractmethod
+    def _build_single_net(self, *, device, **kwargs):
```
[the remark this answers is not in the document]
Comment (vmoens) [reply torchrl/modules/models/multiagent.py:69]: yep that is the idea!

**Comment by kfu02** [comment]
> @kfu02 wanna give a shot at MA-RNNs with this once it's merged? Or should I draft a PR?

Yes! I will put up a draft by the end of the week!

**Comment by matteobettini** [comment]
I ran some benchmarks on the mappo_ippo example for the MLP. With MAPPO and non-sharinmg params for 3 agents.

It seems to work! Do not see any regression in this case

W B Chart 21_02_2024, 11_18_37
W B Chart 21_02_2024, 11_18_50

**Comment by matteobettini** [comment]
Further to #1957 i ran some tests with MASAC and non-shared parameters on the same task for 4 agents and the results look good!
W B Chart 26_02_2024, 09_16_20
W B Chart 26_02_2024, 09_16_32

Almost half training time!
